bokeh/core/_templates/autoload_nb_js.js

```diff
@@ -1,5 +1,116 @@
 {% extends "autoload_js.js" %}
 
+{% block register_mimetype %}
+
+  var MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';
+  var CLASS_NAME = 'output_bokeh rendered_html';
+
+  /**
+   * Render data to the DOM node
+   */
+  function render(props, node) {
+    if (props.data['div'] !== undefined) {
+      var div = document.createElement("div");
+      div.innerHTML = props.data["div"];
+      node.appendChild(div);
+    }
+
+    if (props.data['script'] !== undefined) {
+      var script = document.createElement("script");
+      script.textContent = props.data["script"];
+      node.appendChild(script);
+    }
+  }
+
+  /**
+   * Handle when an output is cleared or removed
+   */
+  function handleClearOutput(event, handle) {
+    var cell = handle.cell;
+
+    var id = cell.output_area._bokeh_element_id;
+    var server_id = cell.output_area._server_id;
+    // Clean up Bokeh references
+    if (id !== undefined) {
+      Bokeh.index[id].remove();
+      delete Bokeh.index[id];
+    }
+
+    if (server_id !== undefined) {
+      // Clean up Bokeh references
+      var cmd = "from bokeh.io import _state; print(_state.uuid_to_server['" + server_id + "'].get_sessions()[0].document.roots[0]._id)";
+      cell.notebook.kernel.execute(cmd, {
+        iopub: {
+          output: function(msg) {
+            var element_id = msg.content.text.trim();
+            Bokeh.index[element_id].remove();
+            delete Bokeh.index[element_id];
+          }
+        }
+      });
+      // Destroy server and session
+      var cmd = "from bokeh import io; io._destroy_server('" + server_id + "')";
+      cell.notebook.kernel.execute(cmd);
+    }
+  }
+
+  /**
+   * Handle when a new output is added
+   */
+  function handleAddOutput(event,  handle) {
+    var output_area = handle.output_area;
+    var output = handle.output;
+    // store reference to embed id or server id on output_area
+    output_area._bokeh_element_id = output.metadata[MIME_TYPE]["id"];
```

Comment:
The rationale for attaching it to the output area is that that is exactly what triggers the notification when it is added/removed, so all of the necessary information is collected locally where it is needed. Keeping it in the module would necessitate some more complicated mapping between all output areas and their respective corresponding server id's which seems more complicated for no gain.